proposals/0000-lambda-layout.md

```diff
@@ -0,0 +1,353 @@
+    [some suggested](https://gitlab.haskell.org/ghc/ghc/issues/4359#note_51110)
+    allowing lambda expressions to have multiple clauses. This was not
+    implemented: The most obvious approach of turning `\` into a layout herald
+    had the disadvantage of making some common idioms invalid.
+    - This can be circumvented by introducing a new expression that isn't required
+      to be backwards compatible with existing idioms.
+
+This proposal, then, aims to overcome the shortcomings of lambda expressions
+and allows a new expression to have the same capabilities as function declarations,
+obviating the need for `-XMultiWayIf`, `-XLambdaCase`, and potentially even
+most of function declaration syntax, if a user wishes to use the new expression
+instead (see Example section). This new expression is introduced with the sequence `\ of`. It
+behaves similarly to lambda expressions, except with layout, guards and multiple clauses; for example,
+here is a definition of filter in terms of fix:
+
+```haskell
+filter = fix \of filter' _ []                 -> []
+                 filter' p (x:xs) | p x       -> x : rest
+                                  | otherwise ->     rest
+                   where rest = filter' p xs
+```
+
+The `\of` can be read as "*lambda* comprised *of* the following cases".
+
+By combining the functionality of several features into one feature in
+a way that's consistent with the rest of the language, it enables users who
+wish to use this functionality to work with a simpler and more consistent
+language.
+
+## Proposed Change Specification
+
+A new extension `-XMultiWayLambda` is implemented. Under this extension, a new
+expression is enabled, introduced by the token sequence `\ of`. The whitespace between  `\` and `of`
+is optional and may contain an arbitrary sequence of whitespace characters.
+`\ of` behaves in a way largely similar to `\`, but it is a layout herald, can have multiple
+clauses, and may contain guards (see BNF for details).
+
+Zero clauses are not permitted, as the type of the expression would be ambiguous.
+This means that with `-XEmptyCase`,
+`-XLambdaCase` still has one (albeit rarely used) construct that cannot be
+replaced by a `\ of`-expression without making it slightly longer, though
+where that is acceptable,
+`\case {}` can (even today) be replaced by `\x -> case x of {}`. This shortcoming
+could potentially
+be addressed in a future proposal, for example by adding absurd patterns, to
+provide a more general solution. Note that the `\case {}` construct only works
+for functions of one argument.
+
+Like the existing behavior for alternatives in case- and
+`\case`-expressions, and equations in function declaration syntax, it is
+possible to use `where` clauses within each clause of a `\ of`-expression.
+
+Explicit layout using braces can be used instead of the implicit layout.
+
+Once https://github.com/ghc-proposals/ghc-proposals/pull/155 is being implemented,
```

Comment:
Please let the link point at the merged proposal document, in the master tree, instead.